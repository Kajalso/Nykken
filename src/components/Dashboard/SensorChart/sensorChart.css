.measurementsdata {
  display: grid;
  grid-template-columns: auto auto;
}

label {
  font-size: smaller;
  margin-right: 1em;
}
