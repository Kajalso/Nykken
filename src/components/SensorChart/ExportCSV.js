import React from "react";
import { CSVLink } from "react-csv";

export const ExportCSV = ({ data = [], dataInfo = {} }) => {
  if (!data || !dataInfo) {
    return <pre></pre>;
  }
  //Headers for the CSV-file download
  const headers = [
    { label: "Measurement", key: "measurement" },
    { label: "Timestamp", key: "time_stamp_utc" },
  ];

  //Create CSV-report
  const csvReport = {
    data: data,
    headers: headers,
    filename:
      dataInfo.title +
      "_from_" +
      data[0].time_stamp_utc +
      "_to_" +
      data[data.length - 1].time_stamp_utc +
      ".csv",
  };

  return (
    <div>
      <CSVLink {...csvReport}> Download as CSV </CSVLink>
    </div>
  );
};
