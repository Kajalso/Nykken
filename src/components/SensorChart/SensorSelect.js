import React, { useState } from "react";
import { useAllDataInfo } from "../../api/useAllDataInfo";

export const SensorSelect = ({ onChange }) => {
  const allDataInfo = useAllDataInfo();

  return (
    <select onChange={onChange}>
      {allDataInfo.map((sensor, i) => (
        <option key={i} value={sensor.sensor_id}>
          {sensor.description}
        </option>
      ))}
    </select>
  );
};
