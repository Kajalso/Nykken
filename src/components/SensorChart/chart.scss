$strokeColor: #e3e4e1;

/* Chart styling */
.chart {
  background-color: white;
  padding: 0.5em 0 0.5em 1em;
  border: 1px solid $strokeColor;
  border-radius: 8px;
  display: inline-block;

  svg {
    overflow: visible;
    display: block;
    margin: auto;
  }
}

/* Marks styling */
.mark-line {
  fill: none;
  stroke-width: 3;
  stroke-linejoin: round;
  stroke-linecap: round;
}

.mark-bar {
  stroke-width: 1.5;
  fill-opacity: 0.8;
}

.mark-area {
  fill-opacity: 0.5;
}

.axis-label {
  font-weight: 400;
  fill: #635f5d;
}

.axis-label.y,
.custom-axis-label.y {
  text-anchor: middle;
}

.axis-date {
  text-anchor: middle;
  font-size: 12px;
  font-weight: 500;
  fill: #635f5d;
}

/* Axes styling */

.tick-line {
  stroke: #c0c0bb3f;
}

.tick-text-x {
  text-anchor: middle;
  font-size: 10px;
  fill: #635f5d;
}

.tick-text-y {
  text-anchor: end;
  font-size: x-small;
  fill: #635f5d;
}

.tick-text-y.right {
  text-anchor: start;
}
