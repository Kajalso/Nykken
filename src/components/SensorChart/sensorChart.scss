$secondaryText: #79807c;
$strokeColor: #e3e4e1;

.selects {
  display: flex;
  justify-content: space-between;
}

.sensor-chart {
  a {
    font-size: 12px;
  }
}

.select-time,
.select-granularity {
  position: relative;
  margin-right: 1.5em;
  > * {
    display: inline-block;
  }

  p {
    margin-right: 1em;
  }
}

h3.section-title {
  margin: 0;
}

@media only screen and (min-width: 768px) {
}
