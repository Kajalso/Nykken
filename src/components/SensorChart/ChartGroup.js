import React from "react";

import { LineChart } from "./LineChart/LineChart";
import { BarChart } from "./BarChart/BarChart";

import { useSensorData } from "../../api/useSensorData";


import "./sensorChart.scss";


const barChartIDs = [5, 8, 10, 12];


export const ChartGroup = ({ 
    dataInfo,
    startDateTime,
    endDateTime,
    granularity
 }) => {
    const id = dataInfo.data_identifier;

  // Fetch sensor data and data info
  const sensorData = useSensorData(id, startDateTime, endDateTime, granularity);
 
 return (
    <div className="sensor-chart">
      <h3 className="section-title">{dataInfo.title}</h3>

      {(!sensorData || !sensorData[0]) && (
        <p className="loading">Loading ...</p>
      )}
      {sensorData && sensorData[0] && (
        <>
          {barChartIDs.includes(id) && (
            <BarChart
              data={sensorData}
              dataInfo={dataInfo}
            />
          )}
          {!barChartIDs.includes(id) && (
            <LineChart
              data={sensorData}
              dataInfo={dataInfo}
            />
          )}
        </>
      )}
    </div>
  );
};
