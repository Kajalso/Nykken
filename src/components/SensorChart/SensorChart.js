import React, { useState } from "react";

import { ReactSelect as Select } from "../Select/Select";

import { LineChart } from "./LineChart/LineChart";
import { BarChart } from "./BarChart/BarChart";
import { Button } from "../Button/Button";
import { CustomTimeModal } from "../Modal/CustomTimeModal";

import { useSensorData } from "../../api/useSensorData";

import "./sensorChart.scss";

const exampleDate = "2021-03-01";
const exampleStartTime = "00:00:00";
const exampleEndTime = "00:11:00";

const barChartIDs = [5, 8, 10, 12];

export const SensorChart = ({ id, dataInfo }) => {
  const [customTimeModalIsOpen, setCustomTimeModalIsOpen] = useState(false);

  const closeCustomTimeModal = () => setCustomTimeModalIsOpen(false);

  const [startDate, setStartDate] = useState(exampleDate);
  const [endDate, setEndDate] = useState(exampleDate);

  const [startTime, setStartTime] = useState(exampleStartTime);
  const [endTime, setEndTime] = useState(exampleEndTime);

  const [startDateTime, setStartDateTime] = useState(startDate + startTime);
  const [endDateTime, setEndDateTime] = useState(endDate + endTime);

  // Fetch sensor data and data info
  const sensorData = useSensorData(id, startDateTime, endDateTime);

  const timeOptions = [
    {
      value: "custom",
      label: "Custom",
    },
  ];

  const handleConfirm = (
    startTimeFromInput,
    endTimeFromInput,
    startDateFromInput,
    endDateFromInput
  ) => {
    setStartTime(startTimeFromInput);
    setEndTime(endTimeFromInput);
    setStartDate(startDateFromInput);
    setEndDate(endDateFromInput);

    // Check for correct time format when using Chrome
    if (startTimeFromInput.length < 6) {
      setStartDateTime(startDateFromInput + startTimeFromInput + ":00");
    } else {
      setStartDateTime(startDateFromInput + startTimeFromInput);
    }
    if (endTimeFromInput.length < 6) {
      setEndDateTime(endDateFromInput + endTimeFromInput + ":00");
    } else {
      setEndDateTime(endDateFromInput + endTimeFromInput);
    }
  };
  /*

  const CustomTimeframe = () => {
    return (
      <>
        <div className="date-picker">
          <div className="from">
            <label>From:</label>
            <input
              type="date"
              value={startDateFromInput}
              onChange={(e) => setStartDateFromInput(e.target.value)}
            />
            <input
              type="time"
              value={startTimeFromInput}
              onChange={(e) => setStartTimeFromInput(e.target.value)}
              step="1"
            />
          </div>
          <div className="until">
            <label>Until:</label>
            <input
              type="date"
              value={endDateFromInput}
              onChange={(e) => setEndDateFromInput(e.target.value)}
            />
            <input
              type="time"
              value={endTimeFromInput}
              onChange={(e) => setEndTimeFromInput(e.target.value)}
              step="1"
            />
          </div>
        </div>
        <Button
          text="Fetch data"
          className="fetch-data"
          onClick={handleConfirm}
        />
      </>
    );
  };*/

  return (
    <div className="sensor-chart">
      <>
        <h3 className="section-title">{dataInfo.title}</h3>
        <div className="select-time">
          <p>Time frame:</p>
          <Button
            text="Custom time frame"
            onClick={() => setCustomTimeModalIsOpen(true)}
          />
          <CustomTimeModal
            sensor={dataInfo}
            isOpen={customTimeModalIsOpen}
            handleConfirm={handleConfirm}
            closeModal={closeCustomTimeModal}
          />
          {/*<Select options={timeOptions} />*/}
        </div>
        {(!sensorData || !sensorData[0]) && (
          <p className="loading">Loading ...</p>
        )}
        {sensorData && sensorData[0] && (
          <>
            {barChartIDs.includes(id) && (
              <BarChart data={sensorData} dataInfo={dataInfo} />
            )}
            {!barChartIDs.includes(id) && (
              <LineChart data={sensorData} dataInfo={dataInfo} />
            )}
          </>
        )}
      </>
    </div>
  );
};
