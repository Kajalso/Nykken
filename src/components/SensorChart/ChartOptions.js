import React, { useState, useRef } from "react";

import { ReactSelect as Select } from "../Select/Select";

import { ExportCSV } from "./ExportCSV";

import { LineChart } from "./LineChart/LineChart";
import { BarChart } from "./BarChart/BarChart";

import { CustomTimeModal } from "../Modal/CustomTimeModal";

import { useSensorData } from "../../api/useSensorData";
import { useSessionStorage } from "../../storage/useSessionStorage";
import { useBarChartIDs } from "../../styles/useChartStyles";

import moreIcon from "../../icons/more.svg";

const chartOptions = [
  {
    value: "edit",
    label: "Edit chart",
  },
  {
    value: "download_png",
    label: "Download PNG",
  },
];

export const ChartOptions = ({ sensors, handleConfirm, handleDownloadPNG }) => {
  const id = sensors[0].dataInfo.data_identifier;
  const [customTimeModalIsOpen, setCustomTimeModalIsOpen] = useState(false);

  const componentRef = useRef();

  const closeCustomTimeModal = () => setCustomTimeModalIsOpen(false);

  // Handle change of time frame
  const handleClick = (chartOption) => {
    if (chartOption.value === "edit") {
      setCustomTimeModalIsOpen(true);
    }
    if (chartOption.value === "download_png") {
      handleDownloadPNG();
    }
  };

  return (
    <>
      <div className="more">
        <Select
          className="more-select-container"
          classNamePrefix="more-select"
          options={chartOptions}
          icon={moreIcon}
          onChange={handleClick}
        />
      </div>
      <CustomTimeModal
        sensor={sensors[0].dataInfo}
        isOpen={customTimeModalIsOpen}
        handleConfirm={handleConfirm}
        closeModal={closeCustomTimeModal}
      />
    </>
  );
};
