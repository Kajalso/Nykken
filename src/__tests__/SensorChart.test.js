import React from 'react';
import { render } from '@testing-library/react'

import { SensorChart } from '../components/SensorChart/SensorChart';


//Suppressing warning as suggested by:
const originalError = global.console.error;
beforeAll(() => {
	global.console.error = jest.fn((...args) => {
		if (typeof args[0] === 'string' && args[0].includes('Please upgrade to at least react-dom@16.9.0')) {
	      return
	    }
	    return originalError.call(console, args)
	});
});

afterAll(() => {
  	global.console.error.mockRestore();
});

//Providing props
const dataInfo = {
    data_identifier: 1,
    description: "Temperature",
    sensor_id: 1,
    title: "Temperature",
    unit: "Â°C",
}

const dataInfo1 = {
    description: "Temperature"
}

const id = {id: 1}

describe('renders without crashing and takes snapshot', () => {
    test('renders correctly', async() => {
        render(< SensorChart dataInfo={dataInfo} />);
    });
    test('takes snapshot', async () => {
        const component = render(< SensorChart dataInfo={dataInfo} />);
        expect(component).toMatchSnapshot();
    })
});

describe('displays correct elements', () => {

    test('displays loading while fetching data', async () =>  {
        const { getByTestId } = render( < SensorChart dataInfo={dataInfo}  />)
        const load = getByTestId('loading');
        expect(load).toHaveTextContent("Loading ...");
    })

    /* test('display correct sensor name (Temperature)', () => {
        const { getByText } = render(< LineChart data={data} dataInfo={dI1} />);
        getByText('Temperature');
    })

    test('display correct sensor name (Air Temperature)', () => {
        const { getByText } = render(< LineChart data={data} dataInfo={dI2} />);
        getByText('Air Temperature');
    })

    test('display correct sensor ID', () => {
        const { getByText } = render(< LineChart data={data} dataInfo={dI1} />);
        getByText('Sensor ID: 1');
    }); */
})
