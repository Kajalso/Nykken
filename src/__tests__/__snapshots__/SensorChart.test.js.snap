// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing and takes snapshot takes snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sensor-chart"
      >
        <h3
          class="section-title"
        >
          Temperature
        </h3>
        <div
          class="select-time"
        >
          <p>
            Time frame:
          </p>
          <div
            class="date-picker"
          >
            <div
              class="from"
            >
              <label>
                From:
              </label>
              <input
                type="date"
                value="2021-03-01"
              />
              <input
                step="1"
                type="time"
                value="00:00:00"
              />
            </div>
            <div
              class="until"
            >
              <label>
                Until:
              </label>
              <input
                type="date"
                value="2021-03-01"
              />
              <input
                step="1"
                type="time"
                value="00:11:00"
              />
            </div>
          </div>
          <button
            class="button fetch-data"
          >
            Fetch data
          </button>
        </div>
        <p
          class="loading"
          data-testid="loading"
        >
          Loading ...
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sensor-chart"
    >
      <h3
        class="section-title"
      >
        Temperature
      </h3>
      <div
        class="select-time"
      >
        <p>
          Time frame:
        </p>
        <div
          class="date-picker"
        >
          <div
            class="from"
          >
            <label>
              From:
            </label>
            <input
              type="date"
              value="2021-03-01"
            />
            <input
              step="1"
              type="time"
              value="00:00:00"
            />
          </div>
          <div
            class="until"
          >
            <label>
              Until:
            </label>
            <input
              type="date"
              value="2021-03-01"
            />
            <input
              step="1"
              type="time"
              value="00:11:00"
            />
          </div>
        </div>
        <button
          class="button fetch-data"
        >
          Fetch data
        </button>
      </div>
      <p
        class="loading"
        data-testid="loading"
      >
        Loading ...
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
